xla/service/gpu/gpu_performance_model.h

```diff
@@ -25,19 +25,100 @@ limitations under the License.
 namespace xla {
 namespace gpu {
 
+struct EstimateRunTimeData {
+  int64_t flops;
+  float bytes_written;
+  float elements_out;
+  absl::Duration write_time;
+  absl::Duration exec_time;
+};
+
 class GpuPerformanceModel {
  public:
   struct RunTimes {
     absl::Duration time_unfused;
     absl::Duration time_fused;
   };
+
+  static EstimateRunTimeData EstimateRunTimeForInstruction(
+      const HloInstruction* instr, const GpuHloCostAnalysis* cost_analysis);
+
   static RunTimes EstimateRunTimes(
       const HloInstruction* producer, const GpuHloCostAnalysis* cost_analysis,
       std::vector<HloInstruction*> fused_users = {}, bool multi_output = false);
 
   // Writes estimated execution time to FusionBackendConfig.reification_cost.
   static void RecordEstimatedRunTime(HloInstruction* instruction,
                                      const GpuHloCostAnalysis* cost_analysis);
+  static absl::Duration ComputeTime(const GpuDeviceInfo& gpu_device_info,
+                                    int64_t n_flops, int64_t n_threads);
+
+  static absl::Duration ProducerInputAccessTime(
+      const GpuHloCostAnalysis* cost_analysis,
+      const GpuDeviceInfo& gpu_device_info, const HloInstruction* producer,
+      const HloInstruction* fused_consumer = nullptr);
+};
+
+class GpuPerformanceWithCollectiveModel : public GpuPerformanceModel {
+ public:
+  // Different algorithms that can be used to perform the collective.
+  enum CollectiveAlgo {
+    RING = 0,
+    TREE,
+  };
+
+  // Table for max system bandwidths GB/s for using NCCL's low latency
+  // algorithm. This is used for intra-node estimate.
+  static constexpr std::array<double, 3> low_latency_max_bandwidths = {
+      39.0 /* Volta*/, 87.7 /* Ampere*/, 87.7 /* Hopper*/
+  };
+
+  // Max bandwidth in GB/s for ring low latency 128 algorithm per channel on a
+  // single-node
+  static constexpr std::array<double, 3> per_channel_max_ring_LL128_bandwidths =
+      {
+          20.0 /* Volta */,
+          20.0 /* Ampere */,
+          36.7 /* Hopper */,
+      };
+
+  // Nvlink unidirectional bandwidth for different compute cap. Note this is per
+  // lane bandwidth.
+  static constexpr double sm60_nvlink_bw = 18.0;
+  static constexpr double sm70_nvlink_bw = 20.0;
+  static constexpr double sm80_nvlink_bw = 20.0;
+  static constexpr double sm90_nvlink_bw = 20.0;
+
+  // PCIE bandwidth
+  static constexpr double pci_bw = 12.0;
```

Comment:
Given the number of options (gen 3 - 5, 1 - 16 lanes) it would be good to comment which case does this number describe.

Reply:
added description, it should be for gen 3 16 lanes

Reply:
Sorry for insisting here but it's still not clear:
1) Why is PCIe mentioned at all? As far as I understand it's not about GPU-host communication here but implies that NVLink is based on PCIe signaling.    
2) It seems that many configurations exist: https://en.wikipedia.org/wiki/NVLink, why is exactly gen3 x16 relevant?
3) The number 12 - is it effective measured throughput in GB/s or something else? (nominal would be ~ 16 GB/s)

Reply:
The PCIe logic is needed here since nccl internally determines number of threads to use based on bandwidth and it uses PCIe bandwidth as the differentiator, if the estimated bandwidth is less than PCIe gen 3 x16, it will fallback to use the default protocol which will use memory fences(even though the payload might not go through CPU), therefore nccl will use less threads in this situation to not over-subscribe comm link. gen3x16 is really just a choice made by nccl based on heuristics and it might change in the future.
As for the 12 GB/S, it's an average effective bandwidth of all protocols nccl uses, each protocol has a different discount factor, the most commonly used ones are:
protocol                                    discount factor
default(uses memory fence)  - 100%
low-latency 128                      - ~95%
low-latency 8                          - 50%
So averaging these out gives about 80% hence the 12GB used here.
In general, as discussed above in one of the comments, we are talking with nccl team internally to try to expose all these details through APIs.